agent/pkg/podman/podman.go

```diff
@@ -0,0 +1,358 @@
+package podman
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"os/exec"
+	"os/user"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	dcontainer "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/pkg/stdcopy"
+	"golang.org/x/exp/slices"
+
+	"github.com/determined-ai/determined/agent/internal/container"
+	"github.com/determined-ai/determined/agent/internal/options"
+	"github.com/determined-ai/determined/agent/pkg/docker"
+	"github.com/determined-ai/determined/agent/pkg/events"
+	"github.com/determined-ai/determined/agent/pkg/singularity"
+	"github.com/determined-ai/determined/master/pkg/archive"
+	"github.com/determined-ai/determined/master/pkg/cproto"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/tasks"
+)
+
+var podmanWrapperEntrypoint = path.Join(tasks.RunDir, tasks.SingularityEntrypointWrapperScript)
+
+const (
+	// TODO(DET-9111): Parameterize this by agent ID.
+	agentTmp         = "/tmp/determined/agent"
+	hostNetworking   = "host"
+	bridgeNetworking = "bridge"
+	envFileName      = "envfile"
+	archivesName     = "archives"
+)
+
+// PodmanContainer captures the state of a container.
+type PodmanContainer struct {
+	*singularity.SingularityContainer
+}
+
+// PodmanClient implements ContainerRuntime.
+type PodmanClient struct {
+	*singularity.SingularityClient
+}
+
+// New returns a new podman client, which launches and tracks containers.
+func New(opts options.PodmanOptions) (*PodmanClient, error) {
+	client, err := singularity.New(options.SingularityOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return &PodmanClient{
+		SingularityClient: client,
+	}, nil
+}
+
+// getPullCommand returns the command and arguments to perform the container image pull.
+func (s *PodmanClient) getPullCommand(req docker.PullImage, image string) (string, []string) {
+	args := []string{"pull"}
+	// if req.ForcePull {
+	// 	args = append(args, "--force") // Use 'podman image rm'?
```

Comment:
Doesn't seem to be a pull option for this, but a...

```
podman run
      --pull string                              Pull image before creating ("always"|"missing"|"never") (default "missing")
```

Reply:
If we do a `podman pull`, is there any benefit to using `--pull always` on the subsequent `podman run`? It's hard to say from the docs I have read.

Reply:
I was commenting on the commented out --force which doesn't apply to podman.   

If we just do explicit pulls for podman, then it is doing a force.

So perhaps nothing is needed.

Reply:
That's what I was thinking too.